<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • HilbertCurve</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HilbertCurve</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/HilbertCurve.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/HilbertCurve">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/HilbertCurve/blob/master/vignettes/HilbertCurve.Rmd"><code>vignettes/HilbertCurve.Rmd</code></a></small>
      <div class="hidden name"><code>HilbertCurve.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{Making 2D Hilbert Curve}
-->
<div id="making-2d-hilbert-curve" class="section level1">
<h1 class="hasAnchor">
<a href="#making-2d-hilbert-curve" class="anchor"></a>Making 2D Hilbert Curve</h1>
<p><strong>Author</strong>: Zuguang Gu ( <a href="mailto:z.gu@dkfz.de">z.gu@dkfz.de</a> )</p>
<p><strong>Date</strong>: 2019-09-29</p>
<p><strong>Package version</strong>: 1.15.1</p>
<hr>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="https://en.wikipedia.org/wiki/Hilbert_curve">Hilbert curve</a> is a type of space-filling curves that folds one dimensional axis into a two dimensional space, but still keeps the locality. It has advantages to visualize data with long axis in following two aspects:</p>
<ol style="list-style-type: decimal">
<li>greatly improve resolution of the visualization fron <span class="math inline">\(n\)</span> to <span class="math inline">\(\sqrt{n}\)</span>;</li>
<li>easy to visualize clusters because generally data points in the axis will also be close in the 2D space.</li>
</ol>
<p>This package aims to provide an easy and flexible way to visualize data through Hilbert curve. The implementation and example figures are based on following sources:</p>
<ul>
<li><a href="http://mkweb.bcgsc.ca/hilbert/" class="uri">http://mkweb.bcgsc.ca/hilbert/</a></li>
<li><a href="http://corte.si/posts/code/hilbert/portrait/index.html" class="uri">http://corte.si/posts/code/hilbert/portrait/index.html</a></li>
<li><a href="http://bioconductor.org/packages/devel/bioc/html/HilbertVis.html" class="uri">http://bioconductor.org/packages/devel/bioc/html/HilbertVis.html</a></li>
</ul>
<p>We first load the packages and set seed for random numbers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(HilbertCurve)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(circlize)
<span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">12345</span>)</code></pre></div>
<p>Following plots show Hilbert curves with level 2, 3, 4, 5:</p>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>As shown in the above plots, as level increases, the length of the curve becomes longer and the curve folds more densely. The number of segments (one segment is marked in red) on the Hilbert curve is <code>4^level - 1</code>. If a Hilbert curve with level 11 is used to map to human chromosome 1, the resolution would be <code>249250621/4^11</code> (approximately 59bp per segment).</p>
<div id="locality" class="section level3">
<h3 class="hasAnchor">
<a href="#locality" class="anchor"></a>Locality</h3>
<p><a href="https://en.wikipedia.org/wiki/Hilbert_curve">Hilbert curve</a> folds one-dimensional axis into a two-dimensional space while still preserves the locality of the data points. We visualize this attribute with a Hilbert curve with level 5. In following animation, the point moves with its natural order on the axis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">1024</span>) {
    hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1024</span>, <span class="dt">level =</span> <span class="dv">5</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>, <span class="dt">arrow =</span> <span class="ot">FALSE</span>)
    <span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> i, <span class="dt">np =</span> <span class="ot">NULL</span>, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">size =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"mm"</span>))
}</code></pre></div>
<p>
<video controls loop><source src="move-on-hilbert-curve-.webm"></source></video></p>
<p>Next, we calculate the pairwise distance between data points in the 2D space and visualize it as heatmap. The numbers on x-axis (top) and y-axis (left) illustrate the position of data points in the original 1D axis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(HilbertVis)
pos =<span class="st"> </span>HilbertVis<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/HilbertVis/man/hilbertCurve.html">hilbertCurve</a></span>(<span class="dv">5</span>)
mat =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(pos))
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ComplexHeatmap)

ht =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span>(mat, <span class="dt">name =</span> <span class="st">"dist"</span>, <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, 
    <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, 
    <span class="dt">heatmap_legend_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">title =</span> <span class="st">"euclidean_dist"</span>))
<span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span>(ht, <span class="dt">padding =</span> <span class="kw">unit</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">2</span>), <span class="st">"mm"</span>))
<span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/decorate_heatmap_body.html">decorate_heatmap_body</a></span>(<span class="st">"dist"</span>, {
    <span class="kw">grid.segments</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>), 
          <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>), <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">2</span>))
    <span class="kw">grid.text</span>(<span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">256</span>, <span class="dv">512</span>, <span class="dv">768</span>, <span class="dv">1024</span>)), <span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">256</span>, <span class="dv">512</span>, <span class="dv">768</span>)<span class="op">/</span><span class="dv">1024</span>, <span class="dt">just =</span> <span class="st">"bottom"</span>, 
        <span class="dt">rot =</span> <span class="dv">90</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">10</span>))
    <span class="kw">grid.text</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">256</span>, <span class="dv">512</span>, <span class="dv">768</span>, <span class="dv">1024</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">256</span>, <span class="dv">512</span>, <span class="dv">768</span>, <span class="dv">1024</span>)<span class="op">/</span><span class="dv">1024</span>, <span class="dv">1</span>, <span class="dt">just =</span> <span class="st">"bottom"</span>,
        <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">10</span>))
})</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-5-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Basically, from the heatmap, data points which are close in the 1D axis are still close in the 2D space (illustrated in the areas around diagonals).</p>
<p>Also there are some data points in (e.g, 1 ~ 256) which are close to data points in (768 ~ 1024)(bottom left or top right area in the heatmap), but it is only due to that the curve folds back. This exception can be easily distinguished by adding assistant grid lines or if users are clear with the structure of the curve.</p>
</div>
</div>
<div id="basic-settings" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-settings" class="anchor"></a>Basic settings</h2>
<p>The <strong>HilbertCurve</strong> package provides a rather simple way to display the data in the form of Hilbert curve. It hides all the technical parts and users only need to think that they are adding graphics on a novel axis based on specifying positions.</p>
<p>Generally, customizing a Hilbert curve follows following steps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(...)    <span class="co"># initialize the curve</span>
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ...)        <span class="co"># add points </span>
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, ...)          <span class="co"># add rectangles</span>
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, ...)       <span class="co"># add polygons</span>
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, ...)      <span class="co"># add lines</span>
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, ...)          <span class="co"># add text</span></code></pre></div>
<p><code><a href="../reference/HilbertCurve.html">HilbertCurve()</a></code> is a constructor function and initializes the Hilbert curve. Following example means initializing a Hilbert curve with level 4 which maps data ranging from 1 to 100. The function returns a <code>HilbertCurve</code> class instance and it can be used to add more graphics later.</p>
<p><code>reference</code> argument here is only used to show the structure of the curve.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>By default, the Hilbert curve starts from the bottom left corner. The start position can be specifyed by <code>start_from</code> option and the orientation of the first segment can be set by <code>first_seg</code>:</p>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-8-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The curve can be though as a folded axis. When the coordinate for this folded axis is initialized, low-level graphics can be added by specifying the positions.</p>
<p>There are several ways to specify the “positions” of the data points. The most common way is to construct a <code>IRanges</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(IRanges)
x =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="dv">100</span>, <span class="dv">20</span>))
s =<span class="st"> </span>x[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span><span class="op">*</span><span class="dv">2</span> <span class="op">-</span><span class="st"> </span><span class="dv">1</span>]
e =<span class="st"> </span>x[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span><span class="op">*</span><span class="dv">2</span>]
ir =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span>(s, e)
ir</code></pre></div>
<pre><code>## IRanges object with 10 ranges and 0 metadata columns:
##            start       end     width
##        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]         2        10         9
##    [2]        14        24        11
##    [3]        32        38         7
##    [4]        39        40         2
##    [5]        51        58         8
##    [6]        75        80         6
##    [7]        81        86         6
##    [8]        87        88         2
##    [9]        90        92         3
##   [10]        93        96         4</code></pre>
<p>Here <code>ir</code> contains intervals which are composed by integers. In <a href="#toc_10">later sections</a>, you will see it is also possible to specify positions with numeric values and very large values.</p>
<div id="points" class="section level3">
<h3 class="hasAnchor">
<a href="#points" class="anchor"></a>Points</h3>
<p>There are two modes for adding points. Normally, intervals are always long in the curve and can not be sufficiently represented by single points, thus, by default, a list of e.g. circles compose the intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ir)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The number of circles used to represent the intervals can be controlled by <code>np</code> (number of points per segment). <code>np</code> controls number of tiny segments that split every Hilbert curve segment (e.g. the first horizontal segment in the left bottom in the curve). Following plot is under <code>np = 3</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">16</span>, <span class="dt">level =</span> <span class="dv">2</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>, <span class="dt">title =</span> <span class="st">"np = 3"</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> <span class="dv">1</span>, <span class="dt">x2 =</span> <span class="dv">2</span>, <span class="dt">np =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-11-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Graphic parameters can be set by <code>gp</code>. Note under this mode, the size of points can only be controlled by <code>np</code> argument. To make it more interesting, you can choose different shapes for the points. There are some pre-defined shapes that you can choose from: “circle”, “square”, “triangle”, “hexagon”, “star”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ir, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir))),
    <span class="dt">shape =</span> <span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"circle"</span>, <span class="st">"square"</span>, <span class="st">"triangle"</span>, <span class="st">"hexagon"</span>, <span class="st">"star"</span>), <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir), <span class="dt">replace =</span> <span class="ot">TRUE</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In above figure, you may notice for some points, the color is more light in the ends. This is because the segments represented by these light circles are not fully covered by user’s input intervals, thus, averaging is applied here. E.g. if only the half of the segment represented by the circle is covered by user’s interval and the color for this circle is set to red, then after the averaging, the color would be semi-red (<code>#FF8080</code>, an average between red and white). Averaging is very important when you visualize the genome with zooming. You can find more detailed explanation in the <a href="#toc_14"><strong>Averaging</strong></a> section.</p>
<p>If <code>np</code> is set to 1 or <code>NULL</code>, there will be second mode for adding points that the points are plotted at the center of every interval in <code>ir</code>. In this case, <code>size</code> argument is used to control the size of the points. This mode is useful if you have a lot of small intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ir, <span class="dt">np =</span> <span class="ot">NULL</span>, <span class="dt">size =</span> <span class="kw">unit</span>(<span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir)), <span class="st">"cm"</span>), <span class="dt">pch =</span> <span class="dv">16</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="segments" class="section level3">
<h3 class="hasAnchor">
<a href="#segments" class="anchor"></a>Segments</h3>
<p>Adding segments is quite straightforward. You can assign <code>lwd</code> to a large value to simulate long and twisted rectangles.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, ir, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lwd =</span> <span class="dv">5</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="rectangles" class="section level3">
<h3 class="hasAnchor">
<a href="#rectangles" class="anchor"></a>Rectangles</h3>
<p><code><a href="../reference/hc_rect-dispatch.html">hc_rect()</a></code> fills intervals by a list of squares. The squares always locate at the turning points of the Hilbert curve and the width/height cannot be modified and it is automatically adjusted according to the level of the curve.</p>
<p>There is a difference from <code><a href="../reference/hc_points-dispatch.html">hc_points()</a></code> with setting <code>shape</code> to <code>square</code>. The number of squares per segment is always 2 for <code><a href="../reference/hc_rect-dispatch.html">hc_rect()</a></code> while it can be any number for <code><a href="../reference/hc_points-dispatch.html">hc_points()</a></code>. When the interval is long enough, <code><a href="../reference/hc_rect-dispatch.html">hc_rect()</a></code> can completely fill the surrounding areas.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, ir, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="st">"#FF000080"</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;"></p>
<p>As you can see some rectangles are not full red as well. It is also because of averaging. See <a href="#toc_14"><strong>Averaging</strong></a> sections for explanation.</p>
</div>
<div id="polygons" class="section level3">
<h3 class="hasAnchor">
<a href="#polygons" class="anchor"></a>Polygons</h3>
<p>Basically polygons are quite similar to the rectangles. The major difference is 1. borders can be set for irregular polygons; 2. there is only single color in one polygon while for rectangles, the color can change in the ends of the interval if the corresponding segment is not completely covered by the input interval.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, ir, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="st">"red"</span>, <span class="dt">col =</span> <span class="st">"black"</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="text" class="section level3">
<h3 class="hasAnchor">
<a href="#text" class="anchor"></a>Text</h3>
<p>Text by <code><a href="../reference/hc_text-dispatch.html">hc_text()</a></code> is always added at the center of each interval in <code>ir</code>. Additional settings such as <code>just</code> and <code>rot</code> can also be set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
labels =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(letters, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, ir, <span class="dt">labels =</span> labels, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/IPosRanges-class.html">width</a></span>(ir)<span class="op">*</span><span class="dv">2</span><span class="op">+</span><span class="dv">5</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-17-1.png" width="576" style="display: block; margin: auto;"></p>
<p>When the intervals are long or when there are polygons representing the intervals, it is better to put text in the visual center of the polygons. For this, <code>centered_by</code> argument can be set to <code>polygon</code>. Compare postions of <code>a</code> and <code>b</code> in following plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, ir)
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, ir, <span class="dt">labels =</span> <span class="st">"a"</span>)
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, ir, <span class="dt">labels =</span> <span class="st">"b"</span>, <span class="dt">centered_by =</span> <span class="st">"polygon"</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-18-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This functionality is used to <a href="#toc_22">add chromosome names to the Hilbert curve</a>.</p>
</div>
<div id="combine-low-level-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#combine-low-level-functions" class="anchor"></a>Combine low-level functions</h3>
<p>With combination of these basic low-level graphic functions, complicated graphics can be easily made:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>)
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span>(<span class="dv">1</span>, <span class="dv">100</span>))   <span class="co"># This is an other way to add background line</span>
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, ir, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir), <span class="dt">transparency =</span> <span class="fl">0.8</span>)))
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, ir[<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>)], <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"red"</span>))
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ir, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir))),
    <span class="dt">shape =</span> <span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"circle"</span>, <span class="st">"square"</span>, <span class="st">"triangle"</span>, <span class="st">"hexagon"</span>, <span class="st">"star"</span>), <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir), <span class="dt">replace =</span> <span class="ot">TRUE</span>))
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, ir, <span class="dt">labels =</span> labels, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/IPosRanges-class.html">width</a></span>(ir)<span class="op">*</span><span class="dv">2</span><span class="op">+</span><span class="dv">5</span>, <span class="dt">col =</span> <span class="st">"blue"</span>, <span class="dt">font =</span> <span class="dv">2</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-19-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="non-integer-positions" class="section level2">
<h2 class="hasAnchor">
<a href="#non-integer-positions" class="anchor"></a>Non-integer positions</h2>
<p>It doesn’t matter if your positions are integers or not. Internally, adjustment is automatically applied.</p>
<p>When positions are not integers, you can specify the positions by <code>x1</code> and <code>x2</code>. All low-level graphical funtions accept <code>x1</code> and <code>x2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="fl">0.1</span>, <span class="fl">0.8</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.15</span>, <span class="fl">0.55</span>), <span class="dt">x2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.25</span>, <span class="fl">0.65</span>))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-20-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This is also useful to specify the positions with huge integers which can not be handled by the <code>IRanges</code> class (e.g. the summation of total length of human chromosomes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run when generating the vignette</span>
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000000000000</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> <span class="dv">400000000000</span>, <span class="dt">x2 =</span> <span class="dv">600000000000</span>)</code></pre></div>
<p>Negative positions are allowed as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run when generating the vignette</span>
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="op">-</span><span class="dv">100</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">4</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> <span class="op">-</span><span class="dv">50</span>, <span class="dt">x2 =</span> <span class="dv">50</span>)</code></pre></div>
</div>
<div id="pixel-mode" class="section level2">
<h2 class="hasAnchor">
<a href="#pixel-mode" class="anchor"></a>Pixel mode</h2>
<p>The pixel mode can be thought as a high resolution version of <code><a href="../reference/hc_rect-dispatch.html">hc_rect()</a></code>. When the level is high (e.g. &gt; 10), the whole 2D space will be almost completely filled by the curve and it is impossible to add or visualize e.g. points on the curve. In this case, the ‘pixel’ mode visualizes each tiny ‘segment’ as a pixel and maps values to colors. Internally, the whole plot is represented as an RGB matrix and every time a new layer is added to the plot, the RGB matrix will be updated according to the color overlay. When all the layers are added, normally a PNG figure is generated directly from the RGB matrix. So the Hilbert curve with level 11 will generate a PNG figure with 2048x2048 resolution. This is extremely useful for visualize genomic data. E.g. If we make a Hilbert curve for human chromosome 1 with level 11, then each pixel can represent 60bp (<code>249250621/2048/2048</code>) which is of very high resolution.</p>
<p>Under ‘pixel’ mode, every time a new layer is added, the image is added to the interactive device as a rastered image.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, ir)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Since color is graphic representation to values, there is only one graphic setting <code>col</code>. You can add more than one layers, and just remember to set transparent colors for the additional layers.</p>
<p>To map values to colors, users can try <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2()</a></code> function in <strong>circlize</strong> package to generate a color mapping function. Another advantage of using <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2()</a></code> is it can be used to generate the corresponding color legend.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))
x =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">10</span>, <span class="dv">960</span>, <span class="dt">length =</span> <span class="dv">100</span>)
x1 =<span class="st"> </span>x<span class="co"># start of each interval</span>
x2 =<span class="st"> </span>x <span class="op">+</span><span class="st"> </span><span class="dv">2</span>    <span class="co"># end of each interval</span>
value =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dv">100</span>)  <span class="co"># associated values</span>
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> <span class="dv">750</span>, <span class="dt">x2 =</span> <span class="dv">850</span>, <span class="dt">col =</span> <span class="st">"#00000010"</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-24-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Grid lines can be added to the plot for better distinguishing blocks in the Hilbert curve. The 2D space will be split into <code>2^(grid_line-1)</code> rows and <code>2^(grid_line-1)</code> columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value), <span class="dt">grid_line =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-25-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Borders can be added by setting <code>border</code> to <code>TRUE</code> or a vector of colors to distinguish different intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value), <span class="dt">border =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-26-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Setting <code>border</code> argument in <code>hc_layer</code> directly modifies the RGB matrix. It adds one pixel border for each polygon and sometimes the border dispears when compressing the image. It is more suggested to add borders by <code><a href="../reference/hc_polygon-dispatch.html">hc_polygon()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value))
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-27-1.png" width="576" style="display: block; margin: auto;"></p>
<p>For the low-level graphic functions introduced before, only <code><a href="../reference/hc_polygon-dispatch.html">hc_polygon()</a></code> and <code><a href="../reference/hc_text-dispatch.html">hc_text()</a></code> can work for pixel mode. E.g. <code><a href="../reference/hc_polygon-dispatch.html">hc_polygon()</a></code> draws partitions and <code><a href="../reference/hc_text-dispatch.html">hc_text()</a></code> adds labels for the partitions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value))
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, <span class="dt">x1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">200</span>, <span class="dv">500</span>), <span class="dt">x2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">200</span>, <span class="dv">500</span>, <span class="dv">1000</span>))
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, <span class="dt">x1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">200</span>, <span class="dv">500</span>), <span class="dt">x2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">200</span>, <span class="dv">500</span>, <span class="dv">1000</span>), 
    <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>), <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">20</span>), <span class="dt">centered_by =</span> <span class="st">"polygon"</span>)</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-28-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The Hilbert curve can be save as a PNG figure by <code><a href="../reference/hc_png-HilbertCurve-method.html">hc_png()</a></code> with resolution <code>2^level x 2^level</code>. Note only the curve itself not including the graphics by <code><a href="../reference/hc_polygon-dispatch.html">hc_polygon()</a></code>/<code><a href="../reference/hc_text-dispatch.html">hc_text()</a></code> is exported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run, only for demonstration</span>
<span class="kw"><a href="../reference/hc_png-HilbertCurve-method.html">hc_png</a></span>(hc, <span class="dt">file =</span> <span class="st">"test.png"</span>)</code></pre></div>
<div id="color-overlay" class="section level3">
<h3 class="hasAnchor">
<a href="#color-overlay" class="anchor"></a>Color overlay</h3>
<p>As discussed before, under “pixel” mode, the Hilbert curve is stored in an RGB matrix, every time a new layer is added, the RGB values that correspond to the new input intervals will be updated according to the color and transparency of the new layer.</p>
<p>By default, the color overlay for layers is applied based on the transparency of the new layer. Under “pixel” mode, each pixel is represented by red, green and blue colors for which the value is between 0 and 1. When a new layer with transparency color is added, the overlayed RGB is calculated as follows (<code>r0</code>, <code>g0</code> and <code>b0</code> correspond to the layer which is already added, <code>r</code>, <code>g</code>, <code>b</code> and <code>alpha</code> correspond to the layer which is going to be added). The method is implemented in <code><a href="../reference/default_overlay.html">default_overlay()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">r =<span class="st"> </span>r<span class="op">*</span>alpha <span class="op">+</span><span class="st"> </span>r0<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>alpha)
g =<span class="st"> </span>g<span class="op">*</span>alpha <span class="op">+</span><span class="st"> </span>g0<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>alpha)
b =<span class="st"> </span>b<span class="op">*</span>alpha <span class="op">+</span><span class="st"> </span>b0<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>alpha)</code></pre></div>
<p>Insteading of globally controlling the color overlay, <code><a href="../reference/hc_layer-dispatch.html">hc_layer()</a></code> provides an <code>overlay</code> argument for which users can specificly control the color overlay in the regions that the input invervals cover. In following example, we change the red regions which are overlapped with the blue regions to green.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">300</span>, <span class="dv">750</span>), <span class="dt">x2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">400</span>, <span class="dv">850</span>), <span class="dt">col =</span> <span class="st">"#0000FF20"</span>,
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) { <span class="co"># it's only applied in [300, 400] and [750, 850]</span>
        l =<span class="st"> </span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0) <span class="co"># `is_white` simple tests whether it is the white color</span>
        
        <span class="co"># change the red background to green</span>
        <span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/any.html">any</a></span>(<span class="op">!</span>l)) {
            r0[<span class="op">!</span>l] =<span class="st"> </span><span class="dv">0</span> 
            g0[<span class="op">!</span>l] =<span class="st"> </span><span class="dv">1</span> 
            b0[<span class="op">!</span>l] =<span class="st"> </span><span class="dv">0</span> 
        }

        <span class="co"># overlay #0000FF20 to the background</span>
        <span class="kw"><a href="../reference/default_overlay.html">default_overlay</a></span>(r0, g0, b0, r, g, b, alpha)
    })</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-31-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The self-defined <code>overlay</code> should accept seven arguments for which, <code>r0</code>, <code>g0</code> and <code>b0</code> correspond to RGB channles for the layers which are already added, <code>r</code>, <code>g</code>, <code>b</code> and <code>alpha</code> correspond to the layer which are going to be added. The elements of i.e. <code>r0</code>, <code>g0</code>, <code>b0</code>, <code>r</code>, <code>g</code>, <code>b</code> only correspond to the pixels that are covered by the intervals in the new layer, and <code>overlay</code> function is applied to every interval in the new layer iteratively.</p>
<p>For <code><a href="../reference/default_overlay.html">default_overlay()</a></code>, if any of <code>r0</code>, <code>g0</code>, <code>b0</code> has <code>NA</code>, the color from the new layer is only used. If any of <code>r</code>, <code>g</code>, <code>b</code> has <code>NA</code>, the color from the background is still used. The overlap of the two-layer colors are only calculated when there are no <code>NA</code> values in the colors.</p>
<p>The color overlay can also be applied to continuous colors. If the color mapping is generated by <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">circlize::colorRamp2()</a></code>, <code><a href="https://rdrr.io/pkg/circlize/man/col2value.html">circlize::col2value()</a></code> function can transform from colors back to the original values (it can be thought as the reversed function of <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2()</a></code>). Since the <code>overlay</code> function only provides color information, we can use <code><a href="https://rdrr.io/pkg/circlize/man/col2value.html">col2value()</a></code> to obtain back the original value and use <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2()</a></code> to generate a new set of colors.</p>
<p>In following plot, we change the color theme in the overlapped areas from “white-red” to “white-blue”. To do this, we first select the non-white pixels, transform back to the original values by <code><a href="https://rdrr.io/pkg/circlize/man/col2value.html">col2value()</a></code> and finally change to another color schema.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x1, <span class="dt">x2 =</span> x2, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">300</span>, <span class="dv">750</span>), <span class="dt">x2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">400</span>, <span class="dv">850</span>), <span class="dt">col =</span> <span class="st">"#00000010"</span>,
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
        <span class="co"># non-white pixels</span>
        l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)

        <span class="co"># original value</span>
        v =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/col2value.html">col2value</a></span>(r0[l], g0[l], b0[l], <span class="dt">col_fun =</span> col_fun)

        <span class="co"># new color schema</span>
        col_fun_new =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))
        col_new =<span class="st"> </span><span class="kw">col_fun_new</span>(v, <span class="dt">return_rgb =</span> <span class="ot">TRUE</span>)
        r0[l] =<span class="st"> </span>col_new[, <span class="dv">1</span>]
        g0[l] =<span class="st"> </span>col_new[, <span class="dv">2</span>]
        b0[l] =<span class="st"> </span>col_new[, <span class="dv">3</span>]

        <span class="co"># overlay #00000010 to the background</span>
        <span class="kw"><a href="../reference/default_overlay.html">default_overlay</a></span>(r0, g0, b0, r, g, b, alpha)
    })</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-32-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In above example, actually you can implement such overlay by first splitting the “red” regions into two parts by whether they overlap with the “grey” regions. Then add the not-overlapped regions, the “grey” regions, and overlapped regions as three independent layers.</p>
<pre><code>hc_layer(red regions not ovelap with blue regions, use white-red color schema)
hc_layer(grey regions)
hc_layer(white-blue regions)
</code></pre>
<p>You can see the <code>overlay</code> option provides an easier and more straightforward way to do it.</p>
<p><a href="#toc_20">In later section</a>, we will demonstrate using self-defined overlay method can efficiently visualize the correspondance between H3K36me3 histone mark and gene body.</p>
</div>
</div>
<div id="add-title-and-legends" class="section level2">
<h2 class="hasAnchor">
<a href="#add-title-and-legends" class="anchor"></a>Add title and legends</h2>
<p>Title can be set by the <code>title</code> argument in <code><a href="../reference/HilbertCurve.html">HilbertCurve()</a></code>. Legend can be passed to <code>legend</code> argument in <code><a href="../reference/HilbertCurve.html">HilbertCurve()</a></code> as a <code>grob</code>/<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legends.html">ComplexHeatmap::Legends</a></code> object or a list of these objects. Basically, colors are the main aesthetic mappings to values, thus, <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend()</a></code> function in <strong>ComplexHeatmap</strong> package can be very helpful to create a legend. Also you can consider <code>legendGrob()</code> in <strong>grid</strong> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">value =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir))
col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))
legend1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">by =</span> <span class="fl">0.2</span>), <span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"continuous"</span>)
legend2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#00FF0080"</span>, <span class="st">"#0000FF80"</span>)), 
    <span class="dt">title =</span> <span class="st">"discrete"</span>)

legend =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(legend1, legend2)

hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>, <span class="dt">title =</span> <span class="st">"points"</span>, <span class="dt">legend =</span> legend)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, ir, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">col_fun</span>(value)))
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, ir, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#00FF0020"</span>, <span class="st">"#0000FF20"</span>), <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ir), <span class="dt">replace =</span> <span class="ot">TRUE</span>)))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-33-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Under “pixel” mode, and when the curve is shown on the interactive device as raster image, title and legends can be added as well. But if the plot is exported by <code><a href="../reference/hc_png-HilbertCurve-method.html">hc_png()</a></code>, there will be not title and no legend.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"green"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
breaks =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>, <span class="dt">by =</span> <span class="dv">2</span>)
lgd1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">at =</span> breaks, <span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"style 1"</span>)
lgd2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(breaks), <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">col_fun</span>(<span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(breaks))), <span class="dt">title =</span> <span class="st">"style 2"</span>)
lgd3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"style 3"</span>)

value =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(x))
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">title =</span> <span class="st">"pixel mode"</span>, 
    <span class="dt">legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(lgd1, lgd2, lgd3))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> x, <span class="dt">x2 =</span> x<span class="op">+</span><span class="dv">2</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, <span class="dt">col =</span> <span class="kw">col_fun</span>(value))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-34-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="averaging-models" class="section level2">
<h2 class="hasAnchor">
<a href="#averaging-models" class="anchor"></a>Averaging models</h2>
<p><strong>Note: this “averaging” concept is very important. If you have unexpected plot, try other <code>mean_mode</code> option.</strong></p>
<p>When using points or rectangles under “normal” mode, or using pixels under “pixel” mode, each point/rectangle/pixel corresponds to a small interval. When overlapping these intervals to the segments of Hilbert curve, and when the segment on the curve is not fully covered by the input intervals, proper averaging method should be applied to summarize the value in the corresponding segments.</p>
<p>Depending on different scenarios, <strong>HilbertCurve</strong> provides three metrics for averaging.</p>
<p>The overlapping model is illustrated in following plot. The red line in the bottom represents one segment on the Hilbert curve. Black lines on the top are the parts of input intervals that overlap with the segment. The thick lines indicate the intersected part between the input intervals and the segment.</p>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-35-1.png" width="576" style="display: block; margin: auto;"></p>
<p>For a segment on the curve, <span class="math inline">\(n\)</span> is the number of input intervals which overlap with the segment (it is 3 in the above plot), <span class="math inline">\(w_i\)</span> is the width of the intersected segments (black thick lines), <span class="math inline">\(x_i\)</span> is the value associated with the original intervals.</p>
<p>The <strong>“absolute”</strong> method which is denoted as <span class="math inline">\(v_a\)</span> and is simple calculated as the mean regardless of the width:</p>
<p><span class="math display">\[ v_a = \frac{\sum_i^n{x_i}}{n} \]</span></p>
<p>The <strong>“weighted”</strong> method which is denoted as <span class="math inline">\(v_w\)</span> is the mean weighted by the width of intersections:</p>
<p><span class="math display">\[ v_w = \frac{\sum_i^n{x_iw_i}}{\sum_i^n{w_i}} \]</span></p>
<p>“Absolute” and “weighted” mode should be applied when background information is not of interest. E.g. when summarizing mean methylation in a small segment, since methylation is only related with CpG sites while not other bases, non-CpG background should be ignored.</p>
<p>The <strong>“w0”</strong> mode is the weighted mean between the intersected parts and un-intersected parts:</p>
<p><span class="math display">\[ v_{w0} = \frac{v_wW + v_bW'}{W+W'} \]</span></p>
<p>for which <span class="math inline">\(W\)</span> is sum of width of the intersected parts and <span class="math inline">\(W'\)</span> is the sum of width for the un-intersected parts. <span class="math inline">\(V_b\)</span> is the value corresponding to the background. E.g when averaging colors which are represented as numeric RGB values, the background value is set to 255 which corresponds to white.</p>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>This section shows several interesting examples.</p>
<p>Visualize rainbow colors:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span>(<span class="dv">100</span>)
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">5</span>)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, <span class="dt">x1 =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">99</span>, <span class="dt">x2 =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">100</span>, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> col, <span class="dt">fill =</span> col))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-36-1.png" width="576" style="display: block; margin: auto;"></p>
<p>or use the rectangles:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">1</span>, <span class="dv">100</span>, <span class="dt">level =</span> <span class="dv">5</span>)
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, <span class="dt">x1 =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">99</span>, <span class="dt">x2 =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">100</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> col, <span class="dt">fill =</span> col))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-37-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Reproduce <a href="http://dds.ec/hilvis/iframe.html">the visualization for IPv4 address</a>. The CIDR data is from <a href="http://www.iwik.org/ipcountry/" class="uri">http://www.iwik.org/ipcountry/</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ComplexHeatmap)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(RColorBrewer)

<span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"cidr_list.RData"</span>, <span class="dt">package =</span> <span class="st">"HilbertCurve"</span>))

cidr_list =<span class="st"> </span>cidr_list[<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(cidr_list, length) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]

country =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(cidr_list), <span class="dt">times =</span> <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(cidr_list, length))

ip =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(cidr_list, <span class="st">"r"</span>)

<span class="co"># convert ip address to numbers</span>
mat =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(ip, <span class="st">"</span><span class="ch">\\</span><span class="st">.|/"</span>), as.numeric))))
start =<span class="st"> </span>mat[, <span class="dv">1</span>]<span class="op">*</span><span class="dv">256</span><span class="op">^</span><span class="dv">3</span> <span class="op">+</span><span class="st"> </span>mat[, <span class="dv">2</span>]<span class="op">*</span><span class="dv">256</span><span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>mat[, <span class="dv">3</span>]<span class="op">*</span><span class="dv">256</span> <span class="op">+</span><span class="st"> </span>mat[, <span class="dv">4</span>]
width =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(mat[, <span class="dv">5</span>], <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/strtoi.html">strtoi</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span>, <span class="dv">32</span> <span class="op">-</span><span class="st"> </span>x), <span class="dt">collapse =</span> <span class="st">""</span>), <span class="dt">base =</span> <span class="dv">2</span>))

<span class="co"># top 8 countries</span>
col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/structure.html">structure</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"grey"</span>, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(cidr_list)), <span class="dt">names =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(cidr_list))
top8_rate =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span>(width, country, sum), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>]<span class="op">/</span><span class="dv">256</span><span class="op">^</span><span class="dv">4</span>
top8 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(top8_rate)
col[top8] =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="dv">8</span>, <span class="st">"Set1"</span>)
top8_rate =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(top8_rate<span class="op">*</span><span class="dv">100</span>), <span class="st">"%"</span>)

<span class="co"># this is the part of using HilbertCurve package</span>
lgd =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(top8, <span class="st">"Others"</span>), <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(top8, top8_rate), <span class="st">"Others"</span>),
    <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(col[top8], <span class="st">"grey"</span>)), <span class="dt">title =</span> <span class="st">"Top8 contries"</span>)
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="dv">0</span>, <span class="dv">256</span><span class="op">^</span><span class="dv">4</span>, <span class="dt">start_from =</span> <span class="st">"topleft"</span>, <span class="dt">first_seg =</span> <span class="st">"horizontal"</span>,
    <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="dt">x1 =</span> start, <span class="dt">x2 =</span> start <span class="op">+</span><span class="st"> </span>width, <span class="dt">col =</span> col[country])</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Duration of Chinese dynasties. Data from <a href="http://www.360doc.com/content/11/1123/02/7829717_166641336.shtml" class="uri">http://www.360doc.com/content/11/1123/02/7829717_166641336.shtml</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"chinese_dynasty.RData"</span>, <span class="dt">package =</span> <span class="st">"HilbertCurve"</span>))

detect_os =<span class="st"> </span><span class="cf">function</span>() {
    <span class="cf">if</span> (<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">'w|W'</span>, .Platform<span class="op">$</span>OS.type)) {
        os =<span class="st"> "Windows"</span>
    } <span class="cf">else</span> {
        <span class="cf">if</span> (<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">'darwin'</span>, version<span class="op">$</span>os)) {
            os =<span class="st"> "MacOS"</span>
        } <span class="cf">else</span> {
            os =<span class="st"> "Linux"</span>
        }
    }
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(os)
}
<span class="co"># default font family for Chinese under different OS</span>
fontfamily =<span class="st"> </span><span class="cf">switch</span>(<span class="kw">detect_os</span>(),
    <span class="dt">Windows =</span> <span class="st">"SimSun"</span>,
    <span class="dt">MacOS =</span> <span class="st">"Songti SC"</span>,
    <span class="dt">Linux =</span> <span class="st">"Songti SC"</span>)
<span class="kw"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span>(<span class="st">"chinese_dynasty.png"</span>, <span class="dt">width =</span> <span class="dv">500</span>, <span class="dt">height =</span> <span class="dv">500</span>)
hc =<span class="st"> </span><span class="kw"><a href="../reference/HilbertCurve.html">HilbertCurve</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(chinese_dynasty[[<span class="dv">2</span>]]), <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(chinese_dynasty[[<span class="dv">3</span>]]), 
    <span class="dt">title =</span> <span class="st">"Chinese dynasty (1122 B.C. ~ 1911 A.D.)"</span>, 
    <span class="dt">title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">16</span>, <span class="dt">fontfamily =</span> fontfamily))
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, <span class="dt">x1 =</span> chinese_dynasty[[<span class="dv">2</span>]], <span class="dt">x2 =</span> chinese_dynasty[[<span class="dv">3</span>]], 
    <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(chinese_dynasty), <span class="dt">transparency =</span> <span class="fl">0.2</span>), 
        <span class="dt">lwd =</span> <span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(chinese_dynasty), <span class="dt">min =</span> <span class="dv">1</span>, <span class="dt">max =</span> <span class="dv">10</span>)))
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, <span class="dt">x1 =</span> chinese_dynasty[[<span class="dv">2</span>]], <span class="dt">x2 =</span> chinese_dynasty[[<span class="dv">3</span>]], <span class="dt">labels =</span> chinese_dynasty[[<span class="dv">1</span>]], 
    <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> (chinese_dynasty[[<span class="dv">3</span>]] <span class="op">-</span><span class="st"> </span>chinese_dynasty[[<span class="dv">2</span>]])<span class="op">/</span><span class="dv">500</span> <span class="op">*</span><span class="st"> </span><span class="dv">10</span> <span class="op">+</span><span class="st"> </span><span class="dv">8</span>, 

<span class="dt">year =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="op">-</span><span class="dv">1000</span>, <span class="dv">2000</span>, <span class="dt">by =</span> <span class="dv">100</span>)
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, <span class="dt">x1 =</span> year, <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(year <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="op">-</span>year, <span class="st">"BC"</span>), year), 
    <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>))
<span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span>())</code></pre></div>
<p>
</p>
<center>
<img src="chinese_dynasty.png">
</center>

</div>
<div id="genomichilbertcurve-specific-for-genomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#genomichilbertcurve-specific-for-genomic-data" class="anchor"></a>GenomicHilbertCurve: specific for genomic data</h2>
<p>To make it easy with genomic data which is always stored as a data frame or a <code>GRanges</code> object, the package provides specific class and functions to deal with. <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code> initializes the Hilbert curve by specifying chromosomes or specific background regions. After that, the data frame or <code>GRanges</code> object which contains genomic intervals can be mapped to the curve.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(chr, species, ...)
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(background, ...)</code></pre></div>
<p>Following functions are depatched to the <code>GenomicHilbertCurve</code> class, and the second argument should be a data frame (which is in bed-format) or a <code>GRanges</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, gr, ...)
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, gr, ...)
<span class="kw"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span>(hc, gr, ...)
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, gr, ...)
<span class="kw"><a href="../reference/hc_text-dispatch.html">hc_text</a></span>(hc, gr, ...)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, gr, ...)</code></pre></div>
<p>Following illustrates several examples of making genome-level Hilbert curves.</p>
<div id="genes" class="section level3">
<h3 class="hasAnchor">
<a href="#genes" class="anchor"></a>Genes</h3>
<p>The first example visualizes genes on chromosome 1 (<a href="https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=451688581_FoCuggvGvavyfGhdj9sSGrGlnErE&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=genes&amp;hgta_track=refGene&amp;hgta_table=0&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=">RefSeq genes for human, hg19</a>). Here random colors are used to distinguish different genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicRanges)
<span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"refseq_chr1.RData"</span>, <span class="dt">package =</span> <span class="st">"HilbertCurve"</span>))
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">5</span>, <span class="dt">reference =</span> <span class="ot">TRUE</span>, 
    <span class="dt">reference_gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">"grey"</span>), <span class="dt">arrow =</span> <span class="ot">FALSE</span>)
<span class="kw"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span>(hc, g, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lwd =</span> <span class="dv">6</span>, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(g))))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-43-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="sequence-conservation" class="section level3">
<h3 class="hasAnchor">
<a href="#sequence-conservation" class="anchor"></a>Sequence conservation</h3>
<p>The conserved regions on human chromosome 1 between mouse and human or between zebrafish and human are downloaded from <a href="http://genome.ucsc.edu/cgi-bin/hgTables?hgsid=472144813_MQ4BbmgW37du53Nw6n9SKKy6mb6r&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=compGeno&amp;hgta_track=placentalChainNet&amp;hgta_table=netMm10&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=">UCSC Table Browser</a>. The parameters for downloading are:</p>
<pre><code>clade: Mammal
genome: Human
assembly: Feb. 2009(GRCh37/hg19)
group: Comparative Genomics
= for mouse
track: Placental Chain/Net
table: Mouse Net (netMm10)
= for zebrafish
track: Vertebrate Chain/Net
table: Zebrafish Net(netDanRer7)
</code></pre>
<p>The “net” alignment which is used here shows the best alignment (the alignment algorithm allows longer gaps than traditional affine gap scoring systems) between each part in human genome and other genomes. For a detailed description of the alignment strategy, please go to <a href="http://genome.ucsc.edu/cgi-bin/hgTables?hgsid=472144813_MQ4BbmgW37du53Nw6n9SKKy6mb6r&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=compGeno&amp;hgta_track=placentalChainNet&amp;hgta_table=netMm10&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=">UCSC Table Browser</a>, and click <strong>describe table schema</strong> button.</p>
<p>In following example, the conservation data is mapped to a level 6 Hilbert curve under “normal” mode. Points are used as the graphic and each segment on the Hilbert curve is split by 3 points. In this case, each point on the curve represents approximately 30kb. Only chromosome 1 is visualized.</p>
<p>The point is fully red if the window which is represented by this point is completely covered by the conserved regions, and it is yellow if it is not covered by any region. Colors are interpolated linearly between red and yellow if it is partially covered by the conserved regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># for generating the legend</span>
lgd =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"yellow"</span>, <span class="st">"red"</span>)), 
    <span class="dt">title =</span> <span class="st">"Conservation"</span>,
    <span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.8</span>, <span class="dv">1</span>), 
    <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"0%"</span>, <span class="st">"20%"</span>, <span class="st">"40%"</span>, <span class="st">"60%"</span>, <span class="st">"80%"</span>, <span class="st">"100%"</span>))
chr1_len =<span class="st"> </span><span class="dv">249250621</span></code></pre></div>
<p>Conservation between human and mouse.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"mouse_net.RData"</span>, <span class="dt">package =</span> <span class="st">"HilbertCurve"</span>))
<span class="kw">seqlengths</span>(mouse) =<span class="st"> </span>chr1_len <span class="co"># it is only used to extract the complement</span>
nonmouse =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">gaps</a></span>(mouse); nonmouse =<span class="st"> </span>nonmouse[<span class="kw">strand</span>(nonmouse) <span class="op">==</span><span class="st"> "*"</span>]
gr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(mouse, nonmouse)
col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"red"</span>, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(mouse)), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"yellow"</span>, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(nonmouse)))
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">6</span>, 
    <span class="dt">title =</span> <span class="st">"Conservation between mouse and human on chr1"</span>,
    <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, gr, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="ot">NA</span>, <span class="dt">fill =</span> col))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-45-1.png" width="720" style="display: block; margin: auto;"></p>
<p>Conservation between human and zebrafish.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"zebrafish_net.RData"</span>, <span class="dt">package =</span> <span class="st">"HilbertCurve"</span>))
<span class="kw">seqlengths</span>(zebrafish) =<span class="st"> </span>chr1_len
nonzebrafish =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">gaps</a></span>(zebrafish); nonzebrafish =<span class="st"> </span>nonzebrafish[<span class="kw">strand</span>(nonzebrafish) <span class="op">==</span><span class="st"> "*"</span>]
gr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(zebrafish, nonzebrafish)
col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"red"</span>, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(zebrafish)), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"yellow"</span>, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(nonzebrafish)))
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">6</span>, 
    <span class="dt">title =</span> <span class="st">"Conservation between zebrafish and human on chr1"</span>,
    <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, gr, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="ot">NA</span>, <span class="dt">fill =</span> col))</code></pre></div>
<p><img src="HilbertCurve_files/figure-html/unnamed-chunk-46-1.png" width="720" style="display: block; margin: auto;"></p>
<p>Compareing these two plots, we have following conclusions:</p>
<ol style="list-style-type: decimal">
<li>Although it is already known that mouse is far more orthologically close to human, the Hilber curve illustrates more clearly that almost all parts of human chromosome 1 have similar sequences on mouse genome.</li>
<li>The start and end of the human chromosome 1 is less conserved to mouse compared to other parts in the chromosome. Also there are several less conserved regions inside the human chromosome.</li>
<li>It illustrates how the regions which are conserved to zebrafish spread in the human chromosome and it gives an estimation of the positions and length of the conserved regions.</li>
</ol>
<p>The idea of visualizing sequencing conservation through the Hilbert curve is from <a href="http://mkweb.bcgsc.ca/hilbert/scientificamerican.mhtml">here</a>.</p>
</div>
<div id="gc-percent" class="section level3">
<h3 class="hasAnchor">
<a href="#gc-percent" class="anchor"></a>GC percent</h3>
<p>Following figure visualizes GC percent on chromosome 1, under “normal” mode. The GC content per 1KB window is calculated by <code>hgGcPercent.bin</code> binary which is downloaded from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/" class="uri">http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/</a>. To speed up, we only used chromosome 1 for demonstration.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">pipe</a></span>(<span class="st">"awk '$1==</span><span class="ch">\"</span><span class="st">chr1</span><span class="ch">\"</span><span class="st">' ~/HilbertCurveTest/hg19_gc_percent_window1000.bed"</span>))
col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(df[[<span class="dv">5</span>]], <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.5</span>, <span class="fl">0.9</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"green"</span>, <span class="st">"#FFFFCC"</span>, <span class="st">"red"</span>))
lgd =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"GC percent"</span>,
    <span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">300</span>, <span class="dv">400</span>, <span class="dv">500</span>, <span class="dv">600</span>), 
    <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"30%"</span>, <span class="st">"40%"</span>, <span class="st">"50%"</span>, <span class="st">"60%"</span>))

hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">6</span>, <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_points-dispatch.html">hc_points</a></span>(hc, df, <span class="dt">np =</span> <span class="dv">3</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]), <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]])))
<span class="kw"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">reduce</a></span>(g), <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="st">"#00000020"</span>, <span class="dt">col =</span> <span class="ot">NA</span>))</code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_chr1_points.png">
</center>

<p>Or visualize GC percent on chromosome 1, under “pixel” mode (you may see the difference between “normal” mode and “pixel” mode here):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">reduce</a></span>(g), <span class="dt">col =</span> <span class="st">"#00000020"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_chr1.png">
</center>

<p>Background is not necessary to be the whole chromosome, it can also be self-defined regions. The background can be specified as a <code>GRanges</code> object and pass to <code>background</code> argument. In following example, only the first half part of chromosome 1 is used to initialize the curve.</p>
<p><strong>Note: when using <code>background</code>, one chromosome can only exist in one row in <code>background</code>.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">background =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(<span class="dt">seqnames =</span> <span class="st">"chr1"</span>, <span class="dt">ranges =</span> <span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span>(<span class="dv">1</span>, <span class="kw"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span>(chr1_len<span class="op">/</span><span class="dv">2</span>)))
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">background =</span> background, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">legend =</span> lgd,
    <span class="dt">title =</span> <span class="st">"First half of chromosome 1"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">reduce</a></span>(g), <span class="dt">col =</span> <span class="st">"#00000020"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_half_chr1.png">
</center>

</div>
<div id="histone-modifications" class="section level3">
<h3 class="hasAnchor">
<a href="#histone-modifications" class="anchor"></a>Histone modifications</h3>
<p>Different modifications (e.g. methylation, acetylation) of the histones affect the chromatin-DNA interaction, thus play roles in transcription regulation. Different types of modifications relates to different types of gene regulation. E.g. H3H4me3 modification is found in actively transcribed promoters. ChIP-sequencing (ChIPSeq) utilizes highthroughput DNA sequencing technology to identify status of chromatin states genome-wide, thus it provides a way to study gene regulation in a global aspect.</p>
<p>In this section, we visualize four types of histone modifications (H3K27ac, H3K36me3, H3K4me3, H3K9me3). The ChIPSeq data are downloaded from Roadmap (<a href="http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/" class="uri">http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/</a>, The bed files are alignments of reads, you need to convert to the basepair or window-based coverage before using following R commands). Following files for four histone modifications are used to visualize:</p>
<pre><code>
UCSD.Lung.H3K27ac.STL002.bed
UCSD.Lung.H3K36me3.STL002.bed
UCSD.Lung.H3K4me3.STL002.bed
UCSD.Lung.H3K9me3.STL002.bed
</code></pre>
<p>Since we are going to plot more than one histome marks, the code for plotting are wrapped as a simple function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GetoptLong)
plot_histone_mark =<span class="st"> </span><span class="cf">function</span>(mark) {
    df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">pipe</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"awk '$5&gt;0 &amp;&amp; $1==</span><span class="ch">\"</span><span class="st">chr1</span><span class="ch">\"</span><span class="st">' ~/HilbertCurveTest/UCSD.Lung.@{mark}.STL002.bed"</span>)), 
        <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
    col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(df[[<span class="dv">5</span>]], <span class="fl">0.99</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))
    lgd1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"Intensity"</span>)
    lgd2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(mark, <span class="st">"gene"</span>), <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FF0000"</span>, <span class="st">"#CCCCCC"</span>)),
        <span class="dt">title =</span> <span class="st">"Layer"</span>)

    hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, 
        <span class="dt">title =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"Intensity of @{mark} mark on chr1"</span>), <span class="dt">legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(lgd1, lgd2))
    <span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]))
    <span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">reduce</a></span>(g), <span class="dt">col =</span> <span class="st">"#00000010"</span>)
}</code></pre></div>
<p>In above code, intensities of histone modifications are mapped to a level 9 Hilbert curve so that each pixel approximately represents a window of 950bp. In each plot, there is an additional grey layer which represents the regions of genes (RefSeq genes) (<code>g</code> is from <code><a href="https://rdrr.io/r/base/system.file.html">system.file("extdata", "refseq_chr1.RData", package = "HilbertCurve")</a></code>).</p>
<p>H3K27ac is a mark for active enhancers which is narrow and sharp. The Hilbert curve shows this attributes and also shows enhancers may exist in form of clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_histone_mark</span>(<span class="st">"H3K27ac"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="H3K27ac_chr1.png">
</center>

<p>H3K36me3 is a mark for actively transcribed gene bodies. The Hilbert curve shows that basically most of H3K36me3 modification with high intensity exists as big blocks.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_histone_mark</span>(<span class="st">"H3K36me3"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="H3K36me3_chr1.png">
</center>

<p>H3K4me3 is a mark for TSS of actively transcribed genes which is also narrow and sharp. The Hilbert curve confirms this and also shows the size of marks are quite consistent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_histone_mark</span>(<span class="st">"H3K4me3"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="H3K4me3_chr1.png">
</center>

<p>H3K9me3 is a mark for repressed genes. The Hilbert curve shows that H3K9me3 modications spread over huge regions in the chromosome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_histone_mark</span>(<span class="st">"H3K9me3"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="H3K9me3_chr1.png">
</center>

<p>When there are more than one layers added and users want to see the correspondance between different genomic features, they can define a color overlay function to enhance the visual effects of the overlapping areas. In following example, when regions for H3K36me3 mark are overlapped with gene bodies, the color theme is changed to white-purple.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">pipe</a></span>(<span class="st">"awk '$5&gt;0 &amp;&amp; $1==</span><span class="ch">\"</span><span class="st">chr1</span><span class="ch">\"</span><span class="st">' ~/HilbertCurveTest/UCSD.Lung.H3K36me3.STL002.bed"</span>), 
    <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(df[[<span class="dv">5</span>]], <span class="fl">0.99</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))
col_fun_new =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(df[[<span class="dv">5</span>]], <span class="fl">0.99</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"purple"</span>))
lgd1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"Intensity"</span>)
lgd2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"H3K36me3"</span>, <span class="st">"overlap"</span>, <span class="st">"gene"</span>),
    <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FF0000"</span>, <span class="st">"purple"</span>, <span class="st">"#CCCCCC"</span>)), <span class="dt">title =</span> <span class="st">"Layer"</span>)

<span class="kw"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span>(<span class="st">"H3K36me3_chr1_overlap.png"</span>, <span class="dt">width =</span> <span class="dv">500</span>, <span class="dt">height =</span> <span class="dv">500</span>)
    title =<span class="st"> "Intensity of H3K36me3 mark on chr1"</span>, legend =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(lgd1, lgd2)<span class="er">)</span>
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]))
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">reduce</a></span>(g), <span class="dt">col =</span> <span class="st">"#00000010"</span>,
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
        l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)
        v =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/col2value.html">col2value</a></span>(r0[l], g0[l], b0[l], <span class="dt">col_fun =</span> col_fun)

        col_new =<span class="st"> </span><span class="kw">col_fun_new</span>(v, <span class="dt">return_rgb =</span> <span class="ot">TRUE</span>)
        r0[l] =<span class="st"> </span>col_new[, <span class="dv">1</span>]
        g0[l] =<span class="st"> </span>col_new[, <span class="dv">2</span>]
        b0[l] =<span class="st"> </span>col_new[, <span class="dv">3</span>]

        <span class="kw"><a href="../reference/default_overlay.html">default_overlay</a></span>(r0, g0, b0, r, g, b, alpha)
    })
<span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span>())</code></pre></div>
<p>
</p>
<center>
<img src="H3K36me3_chr1_overlap.png">
</center>

</div>
<div id="methylation" class="section level3">
<h3 class="hasAnchor">
<a href="#methylation" class="anchor"></a>Methylation</h3>
<p>Next figure visualizs methylation on human chromosome 1. Blue corresponds to un-methylation and red corresponds to full-methylation. One thing which is very important here is that methylation is always measured at single CpG sites, so the mode for averaging should be set to <code>absolute</code> in order to sumarize mean methylation values only on CpG sites. Data are from <a href="http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/" class="uri">http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/</a>.</p>
<p>The methylation plot shows quite clear patterns for un-methylated regions which are small and sharp, fully methylated regions and intermediate methylated regions which are represented as large blocks.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">pipe</a></span>(<span class="st">"awk '$1==</span><span class="ch">\"</span><span class="st">chr1</span><span class="ch">\"</span><span class="st">' ~/HilbertCurveTest/UCSD.Lung.Bisulfite-Seq.STL002.bed"</span>), 
    <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
lgd =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html">Legend</a></span>(<span class="dt">col_fun =</span> col_fun, <span class="dt">title =</span> <span class="st">"Methylation"</span>)
hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="st">"chr1"</span>, <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>, <span class="dt">legend =</span> lgd)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="kw">col_fun</span>(df[[<span class="dv">5</span>]]), <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="methylation_chr1.png">
</center>

</div>
<div id="put-multiple-chromosomes-in-one-curve" class="section level3">
<h3 class="hasAnchor">
<a href="#put-multiple-chromosomes-in-one-curve" class="anchor"></a>Put multiple chromosomes in one curve</h3>
<p>We demonstrated how to visualize one single chromosome through Hilbert curve, but sometimes we also want to put all chromosomes into one plot to get a full view of the genome. One solution is to construct a ‘huge fake’ chromosome which merges all real chromosomes. <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code> makes it easy. You can specify multiple chromosomes by sending a vector to <code>chr</code> argument or specify more than one background intervals in <code>background</code> argument.</p>
<p>In following examples, copy number loss and gain for 22 human chromosomes are visualized. Also color for the overlapping regions between layer for gain events and layer for loss events is changed to purple by defining <code>overlay</code> function. Data is from <a href="http://dgv.tcag.ca/dgv/app/downloads">Database of Genomic Variants, CNV map</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>), <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>)
df_gain =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"~/HilbertCurveTest/Stringent.Gain.hg19.2015-02-03.txt"</span>, 
    <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_gain, <span class="dt">col =</span> <span class="st">"red"</span>)
df_loss =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"~/HilbertCurveTest/Stringent.Loss.hg19.2015-02-03.txt"</span>, 
    <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_loss, <span class="dt">col =</span> <span class="st">"green"</span>, <span class="dt">grid_line =</span> <span class="dv">3</span>, <span class="dt">grid_line_col =</span> <span class="st">"grey"</span>,
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
        l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)
        r[l] =<span class="st"> </span><span class="dv">160</span><span class="op">/</span><span class="dv">255</span>
        g[l] =<span class="st"> </span><span class="dv">32</span><span class="op">/</span><span class="dv">255</span>
        b[l] =<span class="st"> </span><span class="dv">240</span><span class="op">/</span><span class="dv">255</span>
        <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(r, g, b)
})</code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes.png">
</center>

<p>A map which shows the chromosome positions in Hilbert curve is very necessary for knowing which part in the curve represents which chromosome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span>(hc)</code></pre></div>
<p>
</p>
<center>
<img src="map_all_chromosomes.png">
</center>

<p>The map can be overlayed to the Hilbert curve for easy correspondance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>), <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_gain, <span class="dt">col =</span> <span class="st">"red"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_loss, <span class="dt">col =</span> <span class="st">"green"</span>, 
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
        l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)
        r[l] =<span class="st"> </span><span class="dv">160</span><span class="op">/</span><span class="dv">255</span>
        g[l] =<span class="st"> </span><span class="dv">32</span><span class="op">/</span><span class="dv">255</span>
        b[l] =<span class="st"> </span><span class="dv">240</span><span class="op">/</span><span class="dv">255</span>
        <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(r, g, b)
})
<span class="kw"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span>(hc, <span class="dt">add =</span> <span class="ot">TRUE</span>, <span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span>(<span class="dv">22</span>, <span class="dt">transparency =</span> <span class="fl">0.8</span>))</code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes_overlay.png">
</center>

<p>If colors of chromosomes defuse the colors in the original Hilbert curve, you can only draw borders of chromosomes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>), <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_gain, <span class="dt">col =</span> <span class="st">"red"</span>)
<span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df_loss, <span class="dt">col =</span> <span class="st">"green"</span>,
    <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
        l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)
        r[l] =<span class="st"> </span><span class="dv">160</span><span class="op">/</span><span class="dv">255</span>
        g[l] =<span class="st"> </span><span class="dv">32</span><span class="op">/</span><span class="dv">255</span>
        b[l] =<span class="st"> </span><span class="dv">240</span><span class="op">/</span><span class="dv">255</span>
        <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(r, g, b)
})
<span class="kw"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span>(hc, <span class="dt">add =</span> <span class="ot">TRUE</span>, <span class="dt">fill =</span> <span class="ot">NA</span>, <span class="dt">border =</span> <span class="st">"grey"</span>)</code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes_border.png">
</center>

<p>If you want to compare several plots simutaneously, you can use <code>grid.layout()</code> to put them together. Remember to set <code>newpage</code> to <code>FALSE</code> in <code><a href="../reference/HilbertCurve.html">HilbertCurve()</a></code> or <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code>. Also note we set <code>labels</code> in <code><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map()</a></code> to define the labels added to each chromosome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plot_curve =<span class="st"> </span><span class="cf">function</span>(column, ...) {
    cm =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/ColorMapping.html">ColorMapping</a></span>(<span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gain"</span>, <span class="st">"loss"</span>, <span class="st">"both"</span>), <span class="dt">color =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"purple"</span>))
    hc =<span class="st"> </span><span class="kw"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>), <span class="dt">level =</span> <span class="dv">9</span>, <span class="dt">mode =</span> <span class="st">"pixel"</span>,
        <span class="dt">title =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"CNV for @{column}"</span>), ...)  ## `qq()` is from the GetoptLong package

    df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"~/HilbertCurveTest/Stringent.Gain.hg19.2015-02-03.txt"</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>,
        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
    df =<span class="st"> </span>df[df[[column]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, , drop =<span class="st"> </span><span class="ot">FALSE</span>]
    <span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="st">"red"</span>)

    df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"~/HilbertCurveTest/Stringent.Loss.hg19.2015-02-03.txt"</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, 
        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
    df =<span class="st"> </span>df[df[[column]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, , drop =<span class="st"> </span><span class="ot">FALSE</span>]
    <span class="kw"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span>(hc, df, <span class="dt">col =</span> <span class="st">"green"</span>, 
        <span class="dt">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {
            l =<span class="st"> </span><span class="op">!</span><span class="kw"><a href="../reference/is_white.html">is_white</a></span>(r0, g0, b0)
            r[l] =<span class="st"> </span><span class="dv">160</span><span class="op">/</span><span class="dv">255</span>
            g[l] =<span class="st"> </span><span class="dv">32</span><span class="op">/</span><span class="dv">255</span>
            b[l] =<span class="st"> </span><span class="dv">240</span><span class="op">/</span><span class="dv">255</span>
            <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(r, g, b)
    })
    <span class="kw"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span>(hc, <span class="dt">labels =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>, <span class="dt">fill =</span> <span class="ot">NA</span>, <span class="dt">border =</span> <span class="st">"grey"</span>)
}

pn =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"African"</span>, <span class="st">"Asian"</span>, <span class="st">"European"</span>)
<span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">layout =</span> <span class="kw">grid.layout</span>(<span class="dt">nrow =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">3</span>)))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(pn)) {
    <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">layout.pos.row =</span> <span class="dv">1</span>, <span class="dt">layout.pos.col =</span> i))
    <span class="kw">plot_curve</span>(pn[i], <span class="dt">newpage =</span> <span class="ot">FALSE</span>)
    <span class="kw">upViewport</span>()
}</code></pre></div>
<p>
</p>
<center>
<img src="cnv_compare.png">
</center>

</div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.2
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
## 
## attached base packages:
##  [1] stats4    parallel  grid      stats     graphics  grDevices utils     datasets  methods  
## [10] base     
## 
## other attached packages:
##  [1] GetoptLong_0.1.7     GenomicRanges_1.36.1 GenomeInfoDb_1.20.0  RColorBrewer_1.1-2  
##  [5] IRanges_2.18.2       S4Vectors_0.22.1     BiocGenerics_0.30.0  ComplexHeatmap_2.1.0
##  [9] HilbertVis_1.42.0    lattice_0.20-38      circlize_0.4.8       HilbertCurve_1.15.1 
## [13] knitr_1.25           markdown_1.1        
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.2             compiler_3.6.1         XVector_0.24.0         bitops_1.0-6          
##  [5] tools_3.6.1            zlibbioc_1.30.0        digest_0.6.21          clue_0.3-57           
##  [9] evaluate_0.14          memoise_1.1.0          png_0.1-7              rlang_0.4.0           
## [13] pkgdown_1.4.1          xfun_0.9               GenomeInfoDbData_1.2.1 cluster_2.1.0         
## [17] stringr_1.4.0          GlobalOptions_0.1.0    desc_1.2.0             fs_1.3.1              
## [21] rprojroot_1.3-2        R6_2.4.0               rmarkdown_1.15         polylabelr_0.1.0      
## [25] magrittr_1.5           backports_1.1.4        htmltools_0.3.6        MASS_7.3-51.4         
## [29] assertthat_0.2.1       shape_1.4.4            colorspace_1.4-1       stringi_1.4.3         
## [33] RCurl_1.95-4.12        rjson_0.2.20           crayon_1.3.4</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#making-2d-hilbert-curve">Making 2D Hilbert Curve</a><ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#basic-settings">Basic settings</a></li>
      <li><a href="#non-integer-positions">Non-integer positions</a></li>
      <li><a href="#pixel-mode">Pixel mode</a></li>
      <li><a href="#add-title-and-legends">Add title and legends</a></li>
      <li><a href="#averaging-models">Averaging models</a></li>
      <li><a href="#examples">Examples</a></li>
      <li><a href="#genomichilbertcurve-specific-for-genomic-data">GenomicHilbertCurve: specific for genomic data</a></li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
